evm/src/cpu/shift.rs

```diff
@@ -0,0 +1,123 @@
+use plonky2::field::extension::Extendable;
+use plonky2::field::packed::PackedField;
+use plonky2::field::types::Field;
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::ext_target::ExtensionTarget;
+
+use crate::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+use crate::cpu::columns::CpuColumnsView;
+use crate::cpu::membus::NUM_GP_CHANNELS;
+use crate::memory::segments::Segment;
+
+pub(crate) fn eval_packed<P: PackedField>(
+    lv: &CpuColumnsView<P>,
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    let is_shift = lv.op.shl + lv.op.shr;
+    let displacement = lv.mem_channels[1]; // holds the shift displacement d
+    let two_exp = lv.mem_channels[2]; // holds 2^d
+
+    // Not needed here; val is the input and we're verifying that output is
+    // val * 2^d (mod 2^256)
+    //let val = lv.mem_channels[0];
+    //let output = lv.mem_channels[NUM_GP_CHANNELS - 1];
+
+    let shift_table_segment = P::Scalar::from_canonical_u64(Segment::ShiftTable as u64);
+
+    let high_limbs_are_zero = lv.general.shift().displacement_high_limbs_are_zero;
+    yield_constr
+        .constraint(is_shift * (high_limbs_are_zero * high_limbs_are_zero - high_limbs_are_zero));
+
+    // Only lookup the shifting factor when displacement is < 2^32
+    yield_constr.constraint(is_shift * (two_exp.used - high_limbs_are_zero));
+    yield_constr.constraint(is_shift * (two_exp.is_read - high_limbs_are_zero));
```

Comment:
Nit: it would be slightly simpler to just `two_exp.is_read` to one regardless of `high_limbs_are_zero`. The value doesn’t matter when `two_exp.used` is zero.
```suggestion
    yield_constr.constraint(is_shift * (two_exp.is_read - P::ONES));
```